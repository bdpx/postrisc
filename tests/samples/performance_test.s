.text
; Simple test program
; 20! factorial compute
.text
    alloc	61
    ldi		%r15, -100
loop_stop_sard:
    srdi	%r13, %r15, 5
    reps.le.d	%r15, %gz, loop_stop_sard

; performance test - long loop
; for(i = 1000000; i>0; i--) DoSome();

    ldi		%r20, 2500000
    ldi		%r15, 20 ; maximum factorial number
    ldi		%r21, 5
loop_stop: (64)
    addi	%r13, %r13, 5
    sub		%r14, %r14, %r55
    cmps.lt.d	%r24, %r14, %r14
    addi	%r13, %r13, 4
    sub		%r14, %r14, %r55
    cmps.lt.d	%r22, %r14, %r14
    addi	%r13, %r13, 33
    srpi	%r14, %r14, %r55, 13
    sub		%r14, %r13, %r21
    srai	%r14, %r14, 7
    reps.gt.d	%r20, %gz, loop_stop
; print loop counter after loop (must be 0)
    write	"%i64(r20) factorials"
    ldi		%r13, 1
    ldi		%r14, 1
start:
    mul		%r13, %r13, %r14
    write	"factorial: %u64(r13)"
    reps.le.d	%r14, %r15, start

    write	"%i64(r14) %i64(r13)"
.end
