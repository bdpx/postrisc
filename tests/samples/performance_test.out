0 factorials
factorial: 1
factorial: 2
factorial: 6
factorial: 24
factorial: 120
factorial: 720
factorial: 5040
factorial: 40320
factorial: 362880
factorial: 3628800
factorial: 39916800
factorial: 479001600
factorial: 6227020800
factorial: 87178291200
factorial: 1307674368000
factorial: 20922789888000
factorial: 355687428096000
factorial: 6402373705728000
factorial: 121645100408832000
factorial: 2432902008176640000
21 2432902008176640000
